compiler/tests/fixtures/static_string.fe

```diff
@@ -0,0 +1,6 @@
+contract Foo:
+    pub def compare_same_static_string():
+        assert "foo" == "foo"
+
+    pub def return_static_string() -> string3:
+        revert # return "foo"
```

Comment:
I'm trying to get the most basic use case to work but none of these worked so far. returning "foo" gave me "CannotMove" which sent me on a journey of implementing a new location `Code` which I'm not sure if it's the right thing to do or not.

The PR is in a miserable state with many loose ends that didn't work out and got me nowhere. Maybe you'll be able to look at it and give me a few hints that get me back on track. If not, I'll take another stab at it next week with a fresh mind.

Reply:
We should be able to avoid introducing a new `Code` location if we map the string to a memory pointer right away.

This is what I'm thinking:

```
my_string: string11  = "my 11 chars' 
self.my_func("hello")
```

becomes:

```
{
  ...
  data "string0" "my 11 chars"
  data "string1" "hello"  

  code {
    ...
    let my_string := datacopym("string0")
    my_func(datacopym("string1"))
  }
}
```

It might be better to go with a `datacopym` function here that takes in the 32 byte name of the data object and copies its content to memory. This would make for more readable Yul code and also seems to make good use of an abstraction provided by Yul. And again, we don't have any code pointers floating around, so we can just remove that location variant.

So the tricky part is letting the mapper know exactly what string it's supposed to be loading. One thing we could do is add an ID to the the node using `Context`. Another approach could be to use the span of the literal as the ID (e.g. `string35-42`). Either should be fine.

The other part is adding the strings to the Yul contract as data objects. This is the kind of thing I'm hoping the `runtime` builder in the mapper module can be used for. However, I'm not realizing that the API might need to be re-engineered a bit since that just provides a set of statements for the code block.

Reply:
This would be good to talk about early next week, it also ties into the contract factory stuff I'm doing.

Reply:
Thanks! Yeah, I was originally trying it without introducing another location type but ran into various errors from the test suite just hanging up without any indication of whats wrong over OutOfGas errors etc and so I decided to explore this location route out of frustration / lack of other options.

However, today I threw all this away again and actually got a few test cases to work.

>So the tricky part is letting the mapper know exactly what string it's supposed to be loading. One thing we could do is add an ID to the the node using Context. Another approach could be to use the span of the literal as the ID (e.g. string35-42). Either should be fine.

>Another approach could be to use the span of the literal as the ID (e.g. string35-42).

If we have that literal multiple times in the code, doesn't that mean we would end up storing it redundant?

Would there anything me wrong with just taking a hash of the string itself. Let's say we have this code:

```
return "The quick brown fox jumps over the lazy dog"
```

This would be stored as:

```
data "4d741b6f1eb29cb2a9b9911c82f56fa8d73b04959d3d9d222895df6c0b28aa15" "The quick brown fox jumps over the lazy dog"
```

(`4d741b6f1eb29cb2a9b9911c82f56fa8d73b04959d3d9d222895df6c0b28aa15` being keccak("The quick brown fox jumps over the lazy dog")

But am I right to assume that these strings are actually limited to 32 byte? If that's true than the example won't fit and it makes me wonder if there would be anything wrong with just using the string itself as the identifier e.g.

```
data "The quick brown fox" "The quick brown fox"
```

Reply:
>data "The quick brown fox" "The quick brown fox"

I added another commit that does exactly that. There's probably a string reason why that isn't a good idea but for now it gets me one step forward to a point where I don't have any hardcoded data points anymore. However, there is still some strangeness going on. I'll make another comment inline.